### DISCUSSION TITLE
Date format in inertia vuejs 3

### DISCUSSION THREAD

**Discussion by Jijiri89** [inertiajs/inertia#1612] [Help (Vue)]
Please, when I return date column from my laravel model into vue3 components, the date format it returns is not what I want but I tried different ways to change it but I couldn't. The format shows is in this format, year - month - day - time plus time zone.

I want to change it to show this to my own prefer format.

Also, is it anyway to speed up the loading time of the inertia vuejs 3 pages? It is like my pages take too long to load.

**Comment by perisicnikola37** [comment]
Hello,

If you want to change the format of the created_at timestamp, you can add the following code inside your `User.php` model or whatever you want:

```
public function getFormattedCreatedAtAttribute(): string
{
    return $this->created_at->format('d M Y');
}
```

This code defines a method called getFormattedCreatedAtAttribute, which retrieves the created_at timestamp and formats it using the 'd M Y' format.
This naming `getFormattedCreatedAtAttribute` follows a specific naming convention in Laravel's Eloquent ORM. The method name is a combination of three parts:

1. `get` - Indicates that this method retrieves a value.
2. `Formatted` - Describes that the retrieved value is formatted.
3. `CreatedAtAttribute` - Identifies the attribute being retrieved, in this case, the created_at timestamp.

For example, if the original timestamp is `2023-07-16 14:30:00`, the formatted result would be `16 Jul 2023`.

Hope I helped :)

**Comment by Jijiri89** [comment]
Thanks so much for your response. I am most grateful, I am going to test it
out and see if it works, I will let you know.
Also, will it work for any date format in the model or this is for any
created_at alone?
Let say I have another column that says date_birth in the model.php.

**Comment by Jijiri89** [comment]
Please I did but still it is not showing up. In my vuejs template, it is
showing this format below in the screenshot.

**Reply by perisicnikola37** [reply]
Don't see any image

**Comment by Jijiri89** [comment]
Sorry about that!

**Comment by jefhar** [comment]
This really doesn't have anything to do with Inertia, but I would argue that a better way of sending a date from Laravel to JavaScript is to:

1. Make sure your model is casting the field to a Carbon date (`created_at` should already be a Carbon object).
2. Pass the date field as a millisecond time integer `$model->created_at->getTimestampMs()` 
3. In JavaScript, create a Date object from the millisecond time integer: `const createdAt = new Date(props.created_at)` and use whatever [format](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#formats_of_tostring_method_return_values) or [formatting library](https://date-fns.org/v2.30.0/docs/format) you need in the presentation layer.

---
> Also, is it anyway to speed up the loading time of the inertia vuejs 3 pages? It is like my pages take too long to load.

Use the [benchmarking](https://laravel.com/docs/10.x/helpers#benchmarking) helper to find out where your page slows down. There are too many variables, including Xdebug, development mode, and slow database queries to answer that question in an unrelated repository thread.

**Comment by ankushmak2** [comment]
have you used carbon date?

**Comment by Twyar** [comment]
I would personally skip formatting in the back-end and just return the datetime value as it is.

Then use `date-fns` or `moment.js` something similar to make a component for me which formats the time in the frontend. 

This is what I use for example:
```javascript
<script setup>
import { format } from 'date-fns';
import { enUS, nl } from 'date-fns/locale';

const locales = { nl, enUS };
defineProps({
  value: {
    type: String,
    required: true,
  },
  defaultFormat: {
    type: String,
    required: false,
    default: 'PPPP',
  },
});
</script>

<template>
  <span>
    {{ format(new Date(value), defaultFormat, {
      locale: locales[$page.props.lang.current.locale]
    }) }}
  </span>
</template>
```
#### Optional
You can set `locale` to whatever you like.

But personally I use [Laravel Localization](https://github.com/mcamara/laravel-localization) on my website to have it in multiple languages.

I then have the following in my `ShareInertiaData.php` Middleware.
```php
<?php

namespace App\Http\Middleware;

use App\Http\Resources\User\UserResource;
use Closure;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\RedirectResponse;
use Illuminate\Http\Request;
use Illuminate\Http\Response;
use Inertia\Inertia;
use Mcamara\LaravelLocalization\Facades\LaravelLocalization;

class ShareInertiaData
{
    /**
     * Share InertiaData
     *
     * @param Request $request
     * @param Closure $next
     *
     * @return \Illuminate\Http\RedirectResponse|Response|JsonResponse
     */
    public function handle(Request $request, Closure $next): RedirectResponse|Response|JsonResponse
    {
        $lang = [
            'current' => [
                'name' => LaravelLocalization::getCurrentLocaleNative(),
                'locale' => LaravelLocalization::getCurrentLocale(),
            ],
        ];
        foreach (LaravelLocalization::getSupportedLocales() as $localeCode => $properties) {
            $lang['routes'][$localeCode] = [
                'route' => LaravelLocalization::getLocalizedURL($localeCode, null, [], false),
                'properties' => $properties,
            ];
        }

        Inertia::share(array_filter([
            'user' => $request->user() ? new UserResource($request->user()) : null,
            'lang' => $lang,
        ]));

        return $next($request);
    }
}
```

This will return the following props to the front-end.
```json
{
   "props":{
      "lang":{
         "current":{
            "name":"English",
            "locale":"en"
         },
         "routes":{
            "en":{
               "route":"https:\/\/domain.com\/en",
               "properties":{
                  "name":"English",
                  "script":"Latn",
                  "native":"English",
                  "regional":"en_GB"
               }
            },
            "nl":{
               "route":"https:\/\/domain.com\/nl",
               "properties":{
                  "name":"Dutch",
                  "script":"Latn",
                  "native":"Nederlands",
                  "regional":"nl_NL"
               }
            }
         }
      }
   }
}
```
Those page props I use to build a language switcher, but mainly use it to retrieve the current language which I use to translate the date for people.

Additionally I use the npm package [laravel-vue-i18n](https://github.com/xiCO2k/laravel-vue-i18n), this will convert all of the files in your `lang/` folder to `.json` files. This allows you to keep all of your translations in your backend, and then simply do something like this for example.
```javascript
{{ $t('validation.password') }}
```

### This is probably all unnecessary in your use case, but I always add it in projects at the start to future proof it in case the website might need translations in the future. It's a very nice way to format your date in my opinion.

**Comment by Jijiri89** [comment]
This is great, thank you so much! I am most grateful

**Comment by Twyar** [comment]
> Also, is it anyway to speed up the loading time of the inertia vuejs 3 pages? It is like my pages take too long to load.

As for this, I use [Laravel Debugbar](https://github.com/barryvdh/laravel-debugbar) which is an amazing package for development.

You'll get an additional console on development which shows all the Queries, Models, Gates, Sessions, Request, etc loaded on your current page.

The `Queries` tab shows you all the database statements and how long they took to execute, based on this you could try to improve the queries.

**Comment by Jijiri89** [comment]
Thanks for the tip!
Please, with inertia vuejs 3, with the deployment to the cpanel, after
running npm run build, is there any thing to do again or that would be all
for me to upload to cpanel.

**Reply by Twyar** [reply]
It depends on your setup.

If you only have a cPanel where you can upload files then yes you have to manually do `npm run build` on your local computer and then upload everything and that should be enough.

However, what you should to do is create a new file that's called `.env.production`, you can simply to `cp .env .env.production`. I'm prette sure that when you do `npm run build` which does `vite build` it will automatically grab the production environments. This way you can set the variables in the `.env.production` to the correct ones such as your `APP_URL`, `DB_*`, `MAIL_*`, etc.

If you have a VPS or something similar I would recommend looking into CI/CD and Pipelines. This can automatically pull and build your code on the server once you merge your latest changes with a `master` branch on your repo. Personally I always use Gitlab for this, but I assume Github Actions is somewhat the same.

This allows you to have a fully automated workflow.

**Comment by Jijiri89** [comment]
Awesome, thanks for the info! It is possible to use github.com to upload my
laravel inertia vuejs 3 to my shared hosting cpanel app?

**Comment by Jijiri89** [comment]
Hello sir
Please I deployed my laravel inertia vuejs 3 project to cpanel but when I
visit the url of the website, I get this error,



 Vite manifest not found at: /home/copkamen/myapp/public/build/manifest.json
Build the production assets

Run npm run build in your deployment script.

   - Asset bundling with Vite
   <https://laravel.com/docs/9.x/vite#running-vite>

Expand vendor frames

   1. 2 vendor frames

**Reply by Twyar** [reply]
Have you done `npm run build` locally ? If so you need to push your build files to the repo.

Standard path is `/public/build`. This path might be in your `.gitignore` file so Git won’t track the changes and allow you to push them to your repo.

Remove the `/public/build` Line from your `.gitignore` File, commit and push the changes. Then build the files and push those. Now you can put those on the cPanel.

**Comment by Jijiri89** [comment]
Thanks for your response. Please, after running npm run build, I just
zipped the project and uploaded it to my cpanel FileMaker. The backend is
working fine but the front-end is what gives me that error.
I did not use github.com to upload the project to cpanel.

**Reply by Twyar** [reply]
If you did `npm run build` It should’ve build tje files in `/public/build`. It should contain a `manifest.json` as well

Are you sure you’re uploading the correct files ?

**Comment by Jijiri89** [comment]
Yes, I did! Please. I am uploading it again to see.

**Comment by Jijiri89** [comment]
Still, I tied but it is not working. Here is the url :
https://copkamenglish.org/

**Comment by Jijiri89** [comment]
This is how I did it. The following screenshot explains it well.
